lightning: let ignore columns be compatible with tidb backend

glorv opened the pull request:
<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
Let lightning config `mydumper.ignore-columns` be compatible with tidb backend. 

Issue Number: close #xxx <!-- REMOVE this line if no issue to close -->

Problem Summary:

### What is changed and how it works?
Let the kv packend generated by encodeLoop contains the column names with ignore columns filtered instead of the raw column names. 

Proposal: [xxx](url) <!-- REMOVE this line if not applicable -->

What's Changed:

How it Works:

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note <!-- bugfixes or new feature need a release note -->

```release-note
None
```

kennytm commented:
(i think you shouldn't add the "needs-cherry-pick-5.1" label, to avoid confusing the bot)

glorv commented:
> (i think you shouldn't add the "needs-cherry-pick-5.1" label, to avoid confusing the bot)

Done. Forgot v5.1 is in br repo.😅

3pointer commented:
/run-br-integration-tests

3pointer commented:
/run-integration-br-tests

glorv commented:
/run-integration-br-test

glorv commented:
/run-integration-br-test

glorv commented:
/run-integration-br-test

glorv commented:
/run-integration-br-test

glorv commented:
/run-integration-br-test

glorv commented:
ping @3pointer  PTAL
/cc @kennytm  PTAL again, as there is quite a bit changes since the last review to fix the compatibility with generated columns 😅

glorv commented:
/run-integration-br-test

glorv commented:
/run-check_dev_2

glorv commented:
/cc @gozssky PTAL

gozssky commented on br/pkg/lightning/backend/tidb/tidb.go:
```diff
@@ -284,22 +287,27 @@ func (enc *tidbEncoder) Encode(logger log.Logger, row []types.Datum, _ int64, co
 	// See: tests/generated_columns/data/gencol.various_types.0.sql this sql has no columns, so encodeLoop will fill the
 	// column permutation with default, thus enc.columnCnt > len(row).
-	if len(row) > enc.columnCnt {
+	if len(row) < enc.columnCnt {
```
Could you please explain why `len(row)` should be greater than or equal to `columnCnt`(which is actual `columenMaxIdx + 1`)?

Suppose a table has 5 columns, namely `a`, `b`, `c`, `d`, `e`. We have a SQL statement in a source file like `insert into t(a, d, c) values(xxx, xxx, xxx)`. Then the `columnPermutation` should be `(0, -1, 2, 1, -1, -1)`, `columnCnt = columnMaxIdx + 1 = 4`.  `len(row) ` is 3 which is smaller than `columnCnt`, so in this scenario, data cannot be imported correctly.

Is my understanding correct？

glorv replied:
@gozssky In this case, the columnMaxIdx should be 2, so columnCnt = 3.

This check is used to handle there are ignored columns. Giving a source file `insert into t(a, d, c, f) values(xxx, xxx, xxx, xx)`, and ignore field `f` in lightning's config, this is still a valid source file. Then len(row) (4) will bigger than columnCnt (3).

BTW, there may be case that row count is smaller than column count if user manually create table schema has extra column with default value, e.g. `updated timestamp default current_timestamp`, then current lightning will result in an error.

gozssky replied:
Oh, I mistook `i` for `idx` 😂.

gozssky commented on br/tests/lightning_duplicate_detection/run.sh:
```diff
@@ -22,7 +22,7 @@ LOG_FILE1="$TEST_DIR/lightning-duplicate-detection1.log"
 
 # let lightning run a bit slow to avoid some table in the first lightning finish too fast.
-export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/lightning/restore/SlowDownImport=sleep(50)"
+export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/lightning/restore/SlowDownImport=sleep(250)"
```
The test `lightning_distributed_import` also need to inscrease the timeout.

gozssky reviewed (approved):
rest LGTM

glorv commented:
/run-integration-br-test

glorv commented:
/run-integration-br-test

glorv commented:
/merge
